### PR TITLE
[OptApp] Restructure 6 - Collective data container

### PR DESCRIPTION
**Pull request by sunethwarna**
**📝 Description**
This adds `CollectiveDataContainer`. It can hold list of different types of `ContainerVariableDataHolders`. All the opertors defined on `ContainerVariableDataHolders` are defined in this as well. It is required for the `OptimizationApplication` in cases, where several model parts with several sensitivities needs to be falltened out for next iteration update calculation. Once the update is calculated, it needs to be stored on the same containers where they originated from. This `CollectiveDataContainer` allows it.

**🆕 Changelog**
Please summarize the changes in one list to generate the changelog:
- Added `CollectiveDataContainer`
- Added tests

### CODE REVIEW

File: applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.h [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+// Application includes
+#include "container_variable_data_holder.h"
+
+namespace Kratos {
+
+///@name Kratos Classes
+///@{
+
+class KRATOS_API(OPTIMIZATION_APPLICATION) CollectiveVariableDataHolder {
```
Comment (matekelemen) [inline applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.h]: docs

I haven't seen `std::variant` getting used in kratos yet, so I think it'd be best to give a detailed description here what problem this class is meant to solve.

Also, the type names are wild in this class :D

File: applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.h [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            ElementPropertiesContainerVariableDataHolderPointerType>;
+
+    KRATOS_CLASS_POINTER_DEFINITION(CollectiveVariableDataHolder);
+
+    ///@}
+    ///@name Life cycle
+    ///#{
+
+    // Default constructor
+    CollectiveVariableDataHolder() = default;
```
Comment (matekelemen) [inline applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.h]:
```suggestion
    CollectiveVariableDataHolder() noexcept = default;
```

File: applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.h [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+    using ConditionContainerVariableDataHolderPointerType =  ContainerVariableDataHolder<ModelPart::ConditionsContainerType, NonHistoricalContainerDataIO>::Pointer;
+
+    using ElementContainerVariableDataHolderPointerType =  ContainerVariableDataHolder<ModelPart::ElementsContainerType, NonHistoricalContainerDataIO>::Pointer;
+
+    using ConditionPropertiesContainerVariableDataHolderPointerType =  ContainerVariableDataHolder<ModelPart::ConditionsContainerType, PropertiesContainerDataIO>::Pointer;
+
+    using ElementPropertiesContainerVariableDataHolderPointerType =  ContainerVariableDataHolder<ModelPart::ElementsContainerType, PropertiesContainerDataIO>::Pointer;
+
+    using ContainerVariableDataHolderPointerVariantType = std::variant<
```
Comment (matekelemen) [inline applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.h]: please come up with more concise names

File: applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.cpp [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+{
+    for (const auto& p_container_variable_data_holder : rContainerVariableDataHolderPointersList){
+        this->AddVariableDataHolder(p_container_variable_data_holder);
+    }
+}
+
+CollectiveVariableDataHolder::CollectiveVariableDataHolder(const CollectiveVariableDataHolder& rOther)
+{
+    for (const auto& p_container_variable_data_holder : rOther.mContainerVariableDataHolderPointersList) {
+        std::visit([&](auto&& v) {
```
Comment (matekelemen) [inline applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.cpp]:
```suggestion
        std::visit([&](const auto& v) {
```

Comment (matekelemen) [reply applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.cpp]: this applies to all other lambdas you're using in `std::visit` as well.

Comment (matekelemen) [reply applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.cpp]: alternatively, you could define an overload that takes a temporary, and then moves it into the correct container.

File: applications/OptimizationApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 5]
```diff
@@ -126,19 +127,56 @@ void  AddCustomUtilitiesToPython(pybind11::module& m)
...
+        .def(py::self -  float())
+        .def(py::self -= float())
+        .def(py::self *  py::self)
+        .def(py::self *= py::self)
+        .def(py::self *  float())
+        .def(py::self *= float())
+        .def(py::self /  py::self)
+        .def(py::self /= py::self)
+        .def(py::self /  float())
+        .def(py::self /= float())
```
Comment (matekelemen) [inline applications/OptimizationApplication/custom_python/add_custom_utilities_to_python.cpp]: I'll go ahead and repeat my concerns about these operations from the previous PR.

File: applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.cpp [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+    CollectiveVariableDataHolder result;
+    for (const auto& p_container_variable_data_holder : mContainerVariableDataHolderPointersList) {
+        std::visit([&](auto&& v) {
+            result.AddVariableDataHolder(v->Clone());
+        }, p_container_variable_data_holder);
+    }
+    return result;
+}
+
+CollectiveVariableDataHolder CollectiveVariableDataHolder::CloneWithDataInitializedToZero()  const
```
Comment (Igarizza) [inline applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.cpp]: Why do you need to clone a data holder and initialize with zero? Why not initialize new dataHolder with zero? Are there additional parameters, info copied?

File: applications/OptimizationApplication/custom_utilities/container_variable_data_holder_utils.cpp [hunk 7]
```diff
@@ -148,6 +148,17 @@ double ContainerVariableDataHolderUtils::NormInf(const ContainerVariableDataHold
     }));
 }
 
+double ContainerVariableDataHolderUtils::NormInf(
```
Comment (Igarizza) [inline applications/OptimizationApplication/custom_utilities/container_variable_data_holder_utils.cpp]: What is the difference between container_variable_data_holder and Eigen::Vector ?

**Review by RezaNajian — DISMISSED** [review]

File: applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.h (hunk 1 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.h]: done :)

File: applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.cpp (hunk 6 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.cpp]: Done :)

File: applications/OptimizationApplication/custom_utilities/container_variable_data_holder_utils.cpp (hunk 7 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/custom_utilities/container_variable_data_holder_utils.cpp]: `Eigen` has its own lazy expression system  which is constructed at compile time using compile time polymorphism. The `ContainerExpression` and `CollectiveExpressions` does the same but at runtime using runtime polymorphism.

**Comment by sunethwarna** [comment]
@matekelemen @RezaNajian @Igarizza This is ready for review.

File: applications/OptimizationApplication/custom_python/add_custom_utilities_to_python.cpp (hunk 5 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/custom_python/add_custom_utilities_to_python.cpp]: It is required, but done as in the `KratosCore::ContainerExpression`.

File: applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.h (hunk 3 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/custom_utilities/container_variable_data_holder/collective_variable_data_holder.h]: done :)

**Review by matekelemen — APPROVED** [review]
👌

**Comment by sunethwarna** [comment]
Thanks @matekelemen

**Pull request merged**